homeassistant/components/light/greenwave.py

```diff
@@ -104,9 +110,10 @@ def update(self):
         import greenwavereality as greenwave
         doc = greenwave.grab_xml(self._host, self.token)
 
-        for device in doc:
-            if device['did'] == self._did:
-                self._state = int(device['state'])
-                self._brightness = greenwave.hass_brightness(device)
-                self._online = greenwave.check_online(device)
-                self._name = device['name']
+        for room in doc:
+            for device in room['device']:
+                if device['did'] == self._did:
+                    self._state = int(device['state'])
+                    self._brightness = greenwave.hass_brightness(device)
+                    self._online = greenwave.check_online(device)
+                    self._name = device['name']
```

Comment:
That is a realy bad runtime. Make a data object like this: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/sensor/yweather.py#L174-L190

On this data object you can map it to a usefuly data format and access without loop.

You code actual:
- Hit the device with 10x bulbs it call the gateway 10x times
- Loop over 1000x